### PR TITLE
base_api.py: Implement operators and document them in cli help

### PR DESCRIPTION
**Pull request by nuclearcat**
Right now for operators we are using "magic" suffix such as __gt, __lt and so on, which is not documented and a bit non-intuitive. We can make cli tool more intuitive with this patch, where it will hide from user this magic, until we find better mechanism to handle operators as mentioned in https://github.com/kernelci/kernelci-api/issues/356

Example tested:

```
kci node find 'created>2023-09-17' 'state=done' 'group=kver'
```

Fixes: https://github.com/kernelci/kernelci-api/issues/356

### CODE REVIEW

**Comment by gctucker** [comment]
> `kci node find 'created>2023-09-17' 'state=done' 'group=kver'`

Out of interest, would this work too?  Since quotes are required anyway, we could also accept spaces around the operators.  I guess it would just require a small tweak in the regex.

```
kci node find "created > 2023-09-17" "state = done" "group = kver"
```

**Comment by gctucker** [comment]
Ah also we need to make sure the previous syntax still works with e.g. `created__ge=2023-09-17` etc.  I think it should by looking at the code, but it's a bit implicit.

**Comment by nuclearcat** [comment]
I forgot that variables might contain special characters too, just \w is not enough, i updated it, but i think still need to verify what kind of values might exist, as even new pattern might not match some special characters
Also now it is handling correctly space, and sure it handles old commands as well. Btw it is gte, not ge, this is what i mean, that mongodb operators are a bit harder to remember, than basic logical operators.

**Comment by nuclearcat** [comment]
```
            # This pattern allows spaces around the operator, e.g.:
            #   'foo = bar', but it create limitations in the name and value
            # name can't end with space and value can't start with space
            #pattern = r'^(.+\S)\s*([!=<>]+)\s*(\S.+)$'
```
I will leave it here and create PR without spaces support, as they add additional limitations on names and values.

File: kernelci/cli/base_api.py [hunk 1]
```diff
@@ -63,12 +64,36 @@ class AttributesCommand(APICommand):
...
+        rval = {}
+        if not attributes:
+            return rval
+        for attribute in attributes:
+            pattern = r'^([\S]+)([!=<>]+)([\S]+)$'
+            match = re.match(pattern, attribute)
+            if match:
+                attribute, operator, value = match.groups()
+                # if operator is >= then append to attribute '__gte' suffix
+                # and so on
```
Comment (gctucker) [inline kernelci/cli/base_api.py:83]: Ideally this should be a docstring in a separate method to just take care of converting the symbols into the API operators.

File: kernelci/cli/base_api.py [hunk 2]
```diff
@@ -63,12 +64,36 @@ class AttributesCommand(APICommand):
...
+                switch = {
+                    '>': '__gt',
+                    '<': '__lt',
+                    '>=': '__gte',
+                    '<=': '__lte',
+                }
+                if operator in switch:
+                    attribute += switch[operator]
+                elif operator != '=':
+                    # raise error
```
Comment (gctucker) [inline kernelci/cli/base_api.py:93]: Maybe this comment could be removed now?  The next line throws the exception so it's pretty clear.

File: kernelci/cli/base_api.py [hunk 3]
```diff
@@ -63,12 +64,36 @@ class AttributesCommand(APICommand):
...
+            "OPERATOR and can be one of: >, <, >=, <=, =",
         },
     ]
 
     @classmethod
     def _split_attributes(cls, attributes):
-        return dict(
-            tuple(attr.split('=')) for attr in attributes
-        ) if attributes else {}
+        rval = {}
```
Comment (gctucker) [inline kernelci/cli/base_api.py:74]: `rval` is not very meaningful, I first thought it was an "r-value" but now I guess you mean it's the returned value?  Maybe `ret` is less ambiguous here.  Still as per my other comment, I think there could be a slightly cleaner implementation as this function is now doing arguably more than one thing.  But that's not really a blocking issue as far as I can tell.

Comment (nuclearcat) [reply kernelci/cli/base_api.py]: Updated to ret

File: kernelci/cli/base_api.py (hunk 2 above)
Comment (nuclearcat) [reply kernelci/cli/base_api.py]: Removed

File: kernelci/cli/base_api.py (hunk 1 above)
Comment (nuclearcat) [reply kernelci/cli/base_api.py]: I added docstring here, i believe it is also temporary as i think sooner or later we will figure out better way than adding suffix

**Comment by nuclearcat** [comment]
I added also verification, for case you mentioned before:
```
./kci node find "group=kver" "group=abc"
Traceback (most recent call last):
  File "/home/nuclearcat/Documents/kernelci-core/./kci", line 29, in <module>
    kernelci.cli.call(args.command, sys.argv[2:])
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/__init__.py", line 53, in call
    _COMMANDS[name](args)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 71, in main
    sub_main("node", globals(), args)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 775, in sub_main
    status = opts.command(configs, opts)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 417, in call
    return func(*args, **kwargs)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 58, in __call__
    return self._api_call(api, configs, args)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 38, in _api_call
    attributes = self._split_attributes(args.attributes)
  File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 99, in _split_attributes
    raise ValueError(f"Attribute {attribute} already exists")
ValueError: Attribute group already exists
```
Because many might not be aware because of our implementation details we cannot set in conditions same attribute name twice.

**Comment by nuclearcat** [comment]
Tested locally:
```
/kci node find 'created>2023-09-01' --limit 1
[
    {
        "id": "64f12e0b8326c545a780b982",
...
        "created": "2023-09-01T00:19:23.219000",
```

**Comment by gctucker** [comment]
Has this been tested with the staging API instance?  I think this got merged too quickly.

@JenySadadia Have you had a chance to look at this?

**Comment by JenySadadia** [comment]
> Has this been tested with the staging API instance? I think this got merged too quickly.
> 
> @JenySadadia Have you had a chance to look at this?

Not yet. I'll test it soon.

**Review by JenySadadia — COMMENTED** [review]
Tested on staging.

Attributes with `=` works without single quotes around.
```
$ ./kci node find name=checkout created=2023-07-17T11:52:31.913000
[{"id": "64b52b7fb5b4a9b8c05e0874", "kind": "node", "name": "checkout", "path": ["checkout"], "group": null, "revision": {"tree": "kernelci", "url": "https://github.com/kernelci/linux.git", "branch": "staging-mainline", "commit": "273cecedf983df3b9cc3fba85b773be154852975", "describe": "staging-mainline-20230717.0", "version": {"version": 6, "patchlevel": 5, "sublevel": null, "extra": "-rc2-1-g273cecedf983", "name": null}}, "parent": null, "state": "done", "result": null, "artifacts": {"tarball": "http://storage.staging.kernelci.org/api/linux-kernelci-staging-mainline-staging-mainline-20230717.0.tar.gz"}, "data": null, "created": "2023-07-17T11:52:31.913000", "updated": "2023-07-17T12:05:29.780000", "timeout": "2023-07-17T12:52:31.894000", "holdoff": "2023-07-17T12:03:49.144000", "owner": "staging.kernelci.org", "user_groups": []}]
```

However, it throws `ValueError` with other operators.
```
$ ./kci node find name=checkout created>2023-07-17T11:52:31.913000
Traceback (most recent call last):
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/./kci", line 29, in <module>
    kernelci.cli.call(args.command, sys.argv[2:])
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/__init__.py", line 53, in call
    _COMMANDS[name](args)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/node.py", line 71, in main
    sub_main("node", globals(), args)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base.py", line 791, in sub_main
    status = opts.command(configs, opts)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base.py", line 417, in call
    return func(*args, **kwargs)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base_api.py", line 58, in __call__
    return self._api_call(api, configs, args)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/node.py", line 38, in _api_call
    attributes = self._split_attributes(args.attributes)
  File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base_api.py", line 103, in _split_attributes
    raise ValueError(f"Invalid attribute {attribute}")
ValueError: Invalid attribute created
```
I'd suggest returning a proper error message instead of `ValueError` in such case @nuclearcat

**Review by JenySadadia — COMMENTED** [review]
We should also enable the support for `!=` operator. But that is not implemented in the API yet.

**Comment by JenySadadia** [comment]
> I added also verification, for case you mentioned before:
> 
> ```
> ./kci node find "group=kver" "group=abc"
> Traceback (most recent call last):
>   File "/home/nuclearcat/Documents/kernelci-core/./kci", line 29, in <module>
>     kernelci.cli.call(args.command, sys.argv[2:])
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/__init__.py", line 53, in call
>     _COMMANDS[name](args)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 71, in main
>     sub_main("node", globals(), args)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 775, in sub_main
>     status = opts.command(configs, opts)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 417, in call
>     return func(*args, **kwargs)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 58, in __call__
>     return self._api_call(api, configs, args)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 38, in _api_call
>     attributes = self._split_attributes(args.attributes)
>   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 99, in _split_attributes
>     raise ValueError(f"Attribute {attribute} already exists")
> ValueError: Attribute group already exists
> ```
> 
> Because many might not be aware because of our implementation details we cannot set in conditions same attribute name twice.

It seems not working when I provided `created` twice.
```
$ ./kci node find name=checkout 'created>2023-07-17T11:53:31.913000' 'created=2023-07-17T11:52:31.913000'
[{"id": "64b52b7fb5b4a9b8c05e0874", "kind": "node", "name": "checkout", "path": ["checkout"], "group": null, "revision": {"tree": "kernelci", "url": "https://github.com/kernelci/linux.git", "branch": "staging-mainline", "commit": "273cecedf983df3b9cc3fba85b773be154852975", "describe": "staging-mainline-20230717.0", "version": {"version": 6, "patchlevel": 5, "sublevel": null, "extra": "-rc2-1-g273cecedf983", "name": null}}, "parent": null, "state": "done", "result": null, "artifacts": {"tarball": "http://storage.staging.kernelci.org/api/linux-kernelci-staging-mainline-staging-mainline-20230717.0.tar.gz"}, "data": null, "created": "2023-07-17T11:52:31.913000", "updated": "2023-07-17T12:05:29.780000", "timeout": "2023-07-17T12:52:31.894000", "holdoff": "2023-07-17T12:03:49.144000", "owner": "staging.kernelci.org", "user_groups": []}]
```

**Comment by gctucker** [comment]
The issue when providing the same attribute name twice is mentioned on https://github.com/kernelci/kernelci-api/issues/356.  Probably it would be good to have a checklist on that issue with things to do / fix before it's resolved.

**Comment by nuclearcat** [comment]
> > I added also verification, for case you mentioned before:
> > ```
> > ./kci node find "group=kver" "group=abc"
> > Traceback (most recent call last):
> >   File "/home/nuclearcat/Documents/kernelci-core/./kci", line 29, in <module>
> >     kernelci.cli.call(args.command, sys.argv[2:])
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/__init__.py", line 53, in call
> >     _COMMANDS[name](args)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 71, in main
> >     sub_main("node", globals(), args)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 775, in sub_main
> >     status = opts.command(configs, opts)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base.py", line 417, in call
> >     return func(*args, **kwargs)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 58, in __call__
> >     return self._api_call(api, configs, args)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/node.py", line 38, in _api_call
> >     attributes = self._split_attributes(args.attributes)
> >   File "/home/nuclearcat/Documents/kernelci-core/kernelci/cli/base_api.py", line 99, in _split_attributes
> >     raise ValueError(f"Attribute {attribute} already exists")
> > ValueError: Attribute group already exists
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because many might not be aware because of our implementation details we cannot set in conditions same attribute name twice.
> 
> It seems not working when I provided `created` twice.
> 
Because of way how conditions is implemented in kernelci-api (suffix in attribute name, __gt and etc), if you put different conditions - it will not be same attribute name anymore.

**Comment by JenySadadia** [comment]
> > However, it throws `ValueError` with other operators.
> > ```
> > $ ./kci node find name=checkout created>2023-07-17T11:52:31.913000
> > Traceback (most recent call last):
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/./kci", line 29, in <module>
> >     kernelci.cli.call(args.command, sys.argv[2:])
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/__init__.py", line 53, in call
> >     _COMMANDS[name](args)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/node.py", line 71, in main
> >     sub_main("node", globals(), args)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base.py", line 791, in sub_main
> >     status = opts.command(configs, opts)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base.py", line 417, in call
> >     return func(*args, **kwargs)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base_api.py", line 58, in __call__
> >     return self._api_call(api, configs, args)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/node.py", line 38, in _api_call
> >     attributes = self._split_attributes(args.attributes)
> >   File "/home/jeny/kernelCI/kernelci-core_env/src/kernelci-core/kernelci/cli/base_api.py", line 103, in _split_attributes
> >     raise ValueError(f"Invalid attribute {attribute}")
> > ValueError: Invalid attribute created
> > ```
> 
> `>`, `<` and possibly `!` will always need to be quoted or escaped so that the shell won't interpret them as operators, but the same is true of any other character that matches a shell operator.

That's true. I just recommended showing a proper error message to the user instead of returning the exception.

**Pull request merged**